### PR TITLE
[FLINK-25315][tests] add new extensions and utils to help the Junit5 …

### PR DESCRIPTION
**Pull request by yuchengxin**
…migration

## What is the purpose of the change

This pull request adds new extensions and utils to help the Junit5 migration.

## Brief change log

- Add `ParameterizedTestExtension` and two annotations to help to migrate the Junit4 parameterized test
- Add `TempDirUtils` to provide some common methods in `org.junit.rules.TemporaryFolder`
- Migrate `LongHashTableTest` to Junit5 as an example to show how to use the  use `ParameterizedTestExtension` 
- Migrate `PackagedProgramUtilsPipelineTest` to Junit5 as an example to show how to `TempDirUtils`

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with @Public(Evolving): no
- The serializers: no
- The runtime per-record code paths (performance sensitive): no
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
- The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? : no

### CODE REVIEW

**Review by Myasuka — CHANGES_REQUESTED** [review]
Thanks for creating this PR! Please take a look at my comments.

I tried another test class `PackagedProgramUtilsPipelineTest` which both have parameterized tests and temp directory usage:

~~~java
@ExtendWith(ParameterizedTestExtension.class)
public class PackagedProgramUtilsPipelineTest {

    @Parameter
    public TestParameter testParameter;

    @TempDir
    private java.nio.file.Path temporaryFolder;
    @Parameters(name = "testParameter-{0}")
    public static Collection<TestParameter> parameters() {
        return Arrays.asList(
                TestParameter.of(
                        DataSetTestProgram.class,
                        pipeline -> ((Plan) pipeline).getExecutionConfig()),
                TestParameter.of(
                        DataStreamTestProgram.class,
                        pipeline -> ((StreamGraph) pipeline).getExecutionConfig()));
    }

~~~

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+                        HierarchyTraversalMode.TOP_DOWN);
+        if (parameterProviders.isEmpty()) {
+            throw new IllegalStateException("Cannot find any parameter provider");
+        }
+        if (parameterProviders.size() > 1) {
+            throw new IllegalStateException("Multiple parameter providers are found");
+        }
+
+        // Get potential test name
+        String testNameTemplate = parameterProviders.get(0).getAnnotation(Parameters.class).name();
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java]: Already modified as you suggested

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+/** The utils contains some methods same as {@link org.junit.rules.TemporaryFolder}. */
+public class TempDirUtils {
+    private static final String TMP_PREFIX = "junit";
+
+    public static File newFolder(Path path) throws IOException {
+        Path tempPath;
+        if (path != null) {
+            tempPath = Files.createTempDirectory(path, TMP_PREFIX, new FileAttribute[0]);
+        } else {
+            tempPath = Files.createTempDirectory(TMP_PREFIX, new FileAttribute[0]);
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java]: Already modified as you suggested

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        /*
+         * Before checking if the paths are absolute paths, check if create() was ever called,
+         * and if it wasn't, throw IllegalStateException.
+         */
+        File root = base.toFile();
+        for (String path : paths) {
+            if (new File(path).isAbsolute()) {
+                throw new IOException("folder path \'" + path + "\' is not a relative path");
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java]: Already modified as you suggested.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        boolean lastMkdirsCallSuccessful = true;
+        for (String path : paths) {
+            relativePath = new File(relativePath, path);
+            file = new File(root, relativePath.getPath());
+
+            lastMkdirsCallSuccessful = file.mkdirs();
+            if (!lastMkdirsCallSuccessful && !file.isDirectory()) {
+                if (file.exists()) {
+                    throw new IOException(
+                            "a file with the path \'" + relativePath.getPath() + "\' exists");
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java]: Already modified as you suggested.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                    "a folder with the path \'" + relativePath.getPath() + "\' already exists");
+        }
+        return file;
+    }
+
+    public static File newFile(Path folder, String fileName) throws IOException {
+        File file = new File(folder.toFile(), fileName);
+        if (!file.createNewFile()) {
+            throw new IOException(
+                    "a file with the name \'" + fileName + "\' already exists in the test folder");
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java]: Has updated all similar places.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+package org.apache.flink.testutils.junit.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+
+/** The utils contains some methods same as {@link org.junit.rules.TemporaryFolder}. */
+public class TempDirUtils {
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java:27]: I have modified PackagedProgramUtilsPipelineTest.java as a example to use TempDirUtils and ParameterizedTestExtension

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/Parameters.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+import org.junit.runners.Parameterized;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * The annotation is used to replace {@link Parameterized.Parameters} for Junit 5 parameterized
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/Parameters.java]: Already modified as you suggested.

[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/Parameters.java]: I have removed junit4's dependencies.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+package org.apache.flink.testutils.junit.utils;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.FileAttribute;
+
+/** The utils contains some methods same as {@link org.junit.rules.TemporaryFolder}. */
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/utils/TempDirUtils.java]: I have removed junit4's dependencies.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * This extension is used to implement parameterized tests for Junit 5 to replace {@link
+ * Parameterized}.
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java]: I have removed junit4's dependencies.

File: flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        private final Object[] parameterValues;
+
+        public FieldInjectingInvocationContext(String testNameTemplate, Object[] parameterValues) {
+            this.testNameTemplate = testNameTemplate;
+            this.parameterValues = parameterValues;
+        }
+
+        @Override
+        public String getDisplayName(int invocationIndex) {
+            if (testNameTemplate.equals("{index}")) {
```
[the remark this answers is not in the document]
Comment (yuchengxin) [reply flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/testutils/junit/extensions/parameterized/ParameterizedTestExtension.java]: I have set this magic string to a private static variable

**Comment by Myasuka** [comment]
@yuchengxin Thanks for the update!
I think this PR looks good. Would you please update the description as you already introduced another `PackagedProgramUtilsPipelineTest.java` test to show how to use `TempDirUtils` and `ParameterizedTestExtension`.

**Comment by yuchengxin** [comment]
> @yuchengxin Thanks for the update! I think this PR looks good. Would you please update the description as you already introduced another `PackagedProgramUtilsPipelineTest.java` test to show how to use `TempDirUtils` and `ParameterizedTestExtension`.

I hive modified the description of this PR

**Comment by yuchengxin** [comment]
@flinkbot run azure

**Comment by Myasuka** [comment]
@flinkbot run azure

**Review by Myasuka — APPROVED** [review]
Thanks for the update, LGTM

**Comment by Myasuka** [comment]
Thanks for the contributions from @yuchengxin and previous work done by @ruanhang1993 !

**Pull request closed without merging**